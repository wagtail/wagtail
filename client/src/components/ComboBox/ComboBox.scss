// Ensure consistent spacing across the whole component.
// With the scrolling and show/hide of the field, correct spacing is critical.
$spacing: theme('spacing.[2.5]');
$spacing-sm: theme('spacing.5');

.w-combobox {
  width: min(400px, 80vw);
  background: $color-white;
  color: $color-input-text;
  border-radius: theme('borderRadius.DEFAULT');
  font-size: theme('fontSize.18');
  box-shadow: theme('boxShadow.md');
  outline: 10px solid transparent;
}

.w-combobox__field {
  padding: $spacing;
  padding-bottom: 0;

  @include media-breakpoint-up(sm) {
    padding: $spacing-sm;
    padding-bottom: 0;
  }
}

.w-combobox [role='combobox'] {
  margin-bottom: $spacing-sm;

  &[disabled] {
    display: none;
  }
}

.w-combobox__menu {
  max-height: min(480px, 70vh);
  overflow-y: scroll;
}

.w-combobox__optgroup {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-auto-flow: column;
  gap: theme('spacing.[0.5]');
  padding: $spacing;
  padding-top: 0;

  @include media-breakpoint-up(sm) {
    width: 400px;
    padding: $spacing-sm;
    padding-top: 0;
  }
}

.w-combobox__optgroup-label {
  @apply w-label-3;
  grid-column: 1 / span 2;
  margin-bottom: $spacing;
  font-size: 1rem;
  font-weight: 700;

  @include media-breakpoint-up(sm) {
    margin-bottom: $spacing-sm;
  }

  @media (forced-colors: active) {
    color: GrayText;
  }
}

.w-combobox__option {
  display: grid;
  grid-template-columns: theme('spacing.8') 1fr;
  align-items: center;
  padding: theme('spacing.1');
  border: 1px solid transparent;
  font-size: 0.875rem;
  line-height: theme('lineHeight.tight');
  border-radius: theme('borderRadius.sm');

  &[aria-selected='true'] {
    border-color: currentColor;
    background: transparent;
    cursor: pointer;

    @media (forced-colors: active) {
      background: Highlight;
      color: HighlightText;
    }
  }
}

.w-combobox__option--col1 {
  grid-column: 1 / span 1;
}

.w-combobox__option--col2 {
  grid-column: 2 / span 1;
}

.w-combobox__option-icon {
  color: theme('colors.grey.200');
  height: theme('spacing.4');

  .icon {
    width: theme('spacing.4');
    height: theme('spacing.4');
  }

  // Give more width to icons with wide visuals.
  .icon-h1,
  .icon-h2,
  .icon-h3,
  .icon-h4,
  .icon-h5,
  .icon-h6 {
    width: theme('spacing.6');
  }

  // Explicitly override the selected color for SVG support.
  [aria-selected='true'] & {
    @media (forced-colors: active) {
      color: inherit;
    }
  }
}

.w-combobox__option-text {
  // Force to CanvasText even when highlighted, because the extra div
  // makes WHCM add a mandatory Canvas background below the text.
  @media (forced-colors: active) {
    color: CanvasText;
  }
}

.w-combobox__status {
  padding: $spacing-sm;

  @media (forced-colors: active) {
    color: GrayText;
  }
}
