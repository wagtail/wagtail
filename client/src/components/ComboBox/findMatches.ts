// Language-sensitive string comparison.
// See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator/Collator.
const collator = new Intl.Collator(undefined, {
  usage: 'search',
  sensitivity: 'base',
  ignorePunctuation: true,
});

/**
 * Whether a string contains a subsring, with case-insensitive, locale-insensitive search.
 * See https://github.com/adobe/react-spectrum/blob/70e769acf639fc4ef3a704cb8fad81349cb4137a/packages/%40react-aria/i18n/src/useFilter.ts#L57.
 * See also https://github.com/arty-name/locale-index-of,
 * and https://github.com/tc39/ecma402/issues/506.
 */
export const contains = (string: string, substring: string) => {
  if (substring.length === 0) {
    return true;
  }

  const haystack = string.normalize('NFC');
  const needle = substring.normalize('NFC');

  for (let scan = 0; scan + needle.length <= haystack.length; scan += 1) {
    const slice = haystack.slice(scan, scan + needle.length);
    if (collator.compare(needle, slice) === 0) {
      return true;
    }
  }

  return false;
};

/**
 * Find all items where a search field matches the input.
 */
const findMatches = <T extends object>(
  items: T[],
  getSearchFields: (item: T) => (string | null | undefined)[],
  input: string,
) =>
  items.filter((item) => {
    const matches = getSearchFields(item);

    return matches.some((match) => match && contains(match, input));
  });

export default findMatches;
