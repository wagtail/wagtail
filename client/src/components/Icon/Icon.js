import PropTypes from 'prop-types';
import React from 'react';

/**
 * Abstracts away the actual icon implementation (font icons, SVG icons, CSS sprite).
 * Provide a `title` as an accessible label intended for screen readers.
 */
const Icon = ({ name, className, title }) => (
  <span>
    <span className={`icon icon-${name} ${className || ''}`} aria-hidden></span>
    {title ? (
      <span className="visuallyhidden">
        {title}
      </span>
    ) : null}
  </span>
);

Icon.propTypes = {
  name: PropTypes.string.isRequired,
  className: PropTypes.string,
  title: PropTypes.string,
};

Icon.defaultProps = {
  className: null,
  title: null,
};

export default Icon;
