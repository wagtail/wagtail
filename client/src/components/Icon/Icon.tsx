import * as React from 'react';

export interface IconProps {
  name: string;
  className?: string;
  title?: string;
}

/**
 * Provide a `title` as an accessible label intended for screen readers.
 */
const Icon: React.FunctionComponent<IconProps> = ({ name, className, title }) => (
  <>
    <svg className={`icon icon-${name} ${className || ''}`} aria-hidden="true">
      <use href={`#icon-${name}`} />
    </svg>
    {title &&
      <span className="visuallyhidden">
        {title}
      </span>
    }
  </>
);

export default Icon;
