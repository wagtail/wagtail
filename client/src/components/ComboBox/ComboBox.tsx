import React, { useEffect, useState } from 'react';
import {
  useCombobox,
  UseComboboxStateChange,
  UseComboboxStateChangeTypes,
} from 'downshift';

import { gettext } from '../../utils/gettext';
import Icon from '../Icon/Icon';

import findMatches from './findMatches';

export const comboBoxTriggerLabel = gettext('Insert a block');
export const comboBoxLabel = gettext('Search options…');
export const comboBoxNoResults = gettext('No results');

export interface ComboBoxCategory<ItemType> {
  type: string;
  label: string | null;
  items: ItemType[];
}

export interface ComboBoxItem {
  type?: string;
  label?: string | null;
  description?: string | null;
  icon?: string | JSX.Element | null;
  category?: string;
  render?: (props: { option: ComboBoxItem }) => JSX.Element | string;
}

export { UseComboboxStateChange };

export type ComboBoxStateChange = UseComboboxStateChange<ComboBoxItem>;

export interface ComboBoxProps<ComboBoxOption> {
  label?: string;
  placeholder?: string;
  inputValue?: string;
  items: ComboBoxCategory<ComboBoxOption>[];
  getItemLabel: (
    type: string | undefined,
    item: ComboBoxOption,
  ) => string | null | undefined;
  getItemDescription: (item: ComboBoxOption) => string | null | undefined;
  getSearchFields: (item: ComboBoxOption) => (string | null | undefined)[];
  onSelect: (change: UseComboboxStateChange<ComboBoxOption>) => void;
  noResultsText?: string;
}

/**
 * Generic ComboBox component built with downshift, with a 2-column layout.
 */
export default function ComboBox<ComboBoxOption extends ComboBoxItem>({
  label,
  placeholder,
  inputValue,
  items,
  getItemLabel,
  getItemDescription,
  getSearchFields,
  onSelect,
  noResultsText,
}: ComboBoxProps<ComboBoxOption>) {
  // If there is no label defined, we treat the combobox as not needing its own field.
  const inlineCombobox = !label;
  const flatItems = items.flatMap<ComboBoxOption>(
    (category) => category.items || [],
  );
  const [inputItems, setInputItems] = useState<ComboBoxOption[]>(flatItems);
  // Re-create the categories so the two-column layout flows as expected.
  const categories = items.reduce<ComboBoxCategory<ComboBoxOption>[]>(
    (cats, cat, index) => {
      if (cat.label || index === 0) {
        return [...cats, { ...cat, items: cat.items.slice() }];
      }

      // eslint-disable-next-line no-param-reassign
      cats[index - 1].items = cats[index - 1].items.concat(cat.items);

      return cats;
    },
    [],
  );
  const noResults = inputItems.length === 0;
  const {
    getLabelProps,
    getMenuProps,
    getInputProps,
    getItemProps,
    setHighlightedIndex,
    setInputValue,
    openMenu,
  } = useCombobox<ComboBoxOption>({
    ...(typeof inputValue !== 'undefined' && { inputValue }),
    initialInputValue: inputValue || '',
    items: inputItems,
    itemToString(item: ComboBoxOption | null) {
      if (!item) {
        return '';
      }

      return getItemDescription(item) || getItemLabel(item.type, item) || '';
    },
    selectedItem: null,

    onSelectedItemChange: onSelect,

    onInputValueChange: (changes) => {
      const { inputValue: val } = changes;
      if (!val) {
        setInputItems(flatItems);
        return;
      }

      const filtered = findMatches<ComboBoxOption>(
        flatItems,
        getSearchFields,
        val,
      );
      setInputItems(filtered);
      // Always reset the first item to highlighted on filtering, to speed up selection.
      setHighlightedIndex(0);
    },
  });

  useEffect(() => {
    if (inputValue) {
      openMenu();
      setInputValue(inputValue);
      const filtered = findMatches<ComboBoxOption>(
        flatItems,
        getSearchFields,
        inputValue,
      );
      setInputItems(filtered);
      // Always reset the first item to highlighted on filtering, to speed up selection.
      setHighlightedIndex(0);
    } else {
      setInputValue('');
      setInputItems(flatItems);
      setHighlightedIndex(-1);
    }
  }, [inputValue]);

  return (
    <div className="w-combobox">
      {/* downshift does the label-field association itself. */}
      {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
      <label {...getLabelProps()} className="w-sr-only">
        {label}
      </label>
      <div className="w-combobox__field">
        <input
          {...getInputProps()}
          type="text"
          // Prevent the field from receiving focus if it’s not visible.
          disabled={inlineCombobox}
          placeholder={placeholder}
        />
      </div>
      {noResults ? (
        <div className="w-combobox__status">{noResultsText}</div>
      ) : null}
      <div {...getMenuProps()} className="w-combobox__menu">
        {categories.map((category) => {
          const categoryItems = (category.items || []).filter((item) =>
            inputItems.find((i) => i.type === item.type),
          );
          const itemColumns = Math.ceil(categoryItems.length / 2);

          if (categoryItems.length === 0) {
            return null;
          }

          return (
            <div className="w-combobox__optgroup" key={category.type}>
              {category.label ? (
                <div className="w-combobox__optgroup-label">
                  {category.label}
                </div>
              ) : null}
              {categoryItems.map((item, index) => {
                const itemLabel = getItemLabel(item.type, item);
                const description = getItemDescription(item);
                const itemIndex = inputItems.findIndex(
                  (i) => i.type === item.type,
                );
                const itemColumn = index + 1 <= itemColumns ? 1 : 2;
                const hasIcon =
                  typeof item.icon !== 'undefined' && item.icon !== null;
                let icon: JSX.Element | null | undefined = null;

                if (hasIcon) {
                  icon =
                    typeof item.icon === 'string' ? (
                      <Icon name={item.icon} />
                    ) : (
                      item.icon
                    );
                }

                const onMouseDown = (e) => {
                  e.stopPropagation();
                  onSelect({
                    selectedItem: item,
                    type: '__item_click__' as UseComboboxStateChangeTypes.ItemClick,
                  });
                };

                return (
                  // Side-step Downshift event handling and trigger selection on mouse down for clicks,
                  // so we preserve keyboard focus when used within rich text editors.
                  // eslint-disable-next-line jsx-a11y/no-static-element-interactions
                  <div
                    key={item.type}
                    {...getItemProps({ item, index: itemIndex })}
                    onMouseDown={onMouseDown}
                    className={`w-combobox__option w-combobox__option--col${itemColumn}`}
                  >
                    <div className="w-combobox__option-icon">
                      {icon}
                      {/* Support for rich text options using text as an icon (for example "B" for bold). */}
                      {itemLabel && !hasIcon ? <span>{itemLabel}</span> : null}
                    </div>
                    <div className="w-combobox__option-text">
                      {item.render
                        ? item.render({ option: item })
                        : description}
                    </div>
                  </div>
                );
              })}
            </div>
          );
        })}
      </div>
    </div>
  );
}
