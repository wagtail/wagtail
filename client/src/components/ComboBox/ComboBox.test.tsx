import React from 'react';
import { shallow } from 'enzyme';

import ComboBox from './ComboBox';
import Icon from '../Icon/Icon';

const testProps = {
  label: 'Search optionsâ€¦',
  placeholder: 'Search optionsâ€¦',
  getItemLabel: (_, item) => item.label,
  getItemDescription: (item) => item.description,
  getSearchFields: (item) => [item.label, item.description, item.type],
  noResultsText: 'No results, sorry!',
  onSelect: () => {},
};

describe('ComboBox', () => {
  it('renders empty', () => {
    const wrapper = shallow(<ComboBox {...testProps} items={[]} />);
    expect(wrapper.find('.w-combobox__status').text()).toBe(
      'No results, sorry!',
    );
  });

  describe('rendering', () => {
    let items;
    let wrapper;

    beforeEach(() => {
      items = [
        {
          type: 'blockTypes',
          label: 'Blocks',
          items: [
            {
              type: 'blockquote',
              description: 'Blockquote',
              icon: 'blockquote',
            },
            {
              type: 'paragraph',
              description: 'Paragraph',
              icon: <span className="custom-icon">P</span>,
            },
            {
              type: 'heading-one',
              label: 'H1',
              description: 'Heading 1',
            },
            {
              type: 'heading-two',
              label: 'H2',
              render: ({ option }) => (
                <span className="custom-text">{option.label}</span>
              ),
            },
          ],
        },
        {
          type: 'entityTypes',
          items: [
            {
              type: 'link',
              label: 'ðŸ”—',
              description: 'Link',
            },
          ],
        },
      ];
      wrapper = shallow(<ComboBox {...testProps} items={items} />);
    });

    it('shows items', () => {
      const options = wrapper.find('.w-combobox__option-text');
      expect(options).toHaveLength(
        items[0].items.length + items[1].items.length,
      );
      expect(options.at(0).text()).toBe('Blockquote');
    });

    it('uses Icon component', () => {
      expect(wrapper.find(Icon).at(0).prop('name')).toBe('blockquote');
    });

    it('supports custom icons', () => {
      expect(wrapper.find('.custom-icon').text()).toBe('P');
    });

    it('supports label as icon', () => {
      expect(wrapper.find('.custom-text').text()).toBe('H2');
    });

    it('combines two categories into one, with two columns', () => {
      expect(wrapper.find('.w-combobox__optgroup-label')).toHaveLength(1);
      expect(wrapper.find('.w-combobox__option--col1')).toHaveLength(3);
      expect(wrapper.find('.w-combobox__option--col2')).toHaveLength(2);
    });
  });
});
