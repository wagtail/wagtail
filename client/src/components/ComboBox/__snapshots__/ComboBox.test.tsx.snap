// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboBox rendering matches the snapshot 1`] = `
<div
  className="w-combobox-container"
>
  <div
    className="w-combobox"
  >
    <label
      className="w-sr-only"
      htmlFor="downshift-1-input"
      id="downshift-1-label"
    >
      Search optionsâ€¦
    </label>
    <div
      className="w-combobox__field"
    >
      <input
        aria-activedescendant=""
        aria-autocomplete="list"
        aria-controls="downshift-1-menu"
        aria-expanded={false}
        aria-labelledby="downshift-1-label"
        autoComplete="off"
        disabled={false}
        id="downshift-1-input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="Search optionsâ€¦"
        role="combobox"
        type="text"
        value=""
      />
    </div>
    <div
      aria-labelledby="downshift-1-label"
      className="w-combobox__menu"
      id="downshift-1-menu"
      onMouseLeave={[Function]}
      role="listbox"
    >
      <div
        className="w-combobox__optgroup"
        key="blockTypes"
      >
        <div
          className="w-combobox__optgroup-label"
        >
          Blocks
        </div>
        <div
          className="w-combobox__option-row w-combobox__option-row--col1"
          key="blockquote"
        >
          <div
            aria-selected="false"
            className="w-combobox__option"
            id="downshift-1-item-0"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            role="option"
          >
            <div
              className="w-combobox__option-icon"
            >
              <Icon
                name="blockquote"
              />
            </div>
            <div
              className="w-combobox__option-text"
            >
              Blockquote
            </div>
          </div>
        </div>
        <div
          className="w-combobox__option-row w-combobox__option-row--col1"
          key="paragraph"
        >
          <div
            aria-selected="false"
            className="w-combobox__option"
            id="downshift-1-item-1"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            role="option"
          >
            <div
              className="w-combobox__option-icon"
            >
              <span
                className="my-icon"
              >
                P
              </span>
            </div>
            <div
              className="w-combobox__option-text"
            >
              Paragraph
            </div>
          </div>
        </div>
        <div
          className="w-combobox__option-row w-combobox__option-row--col1"
          key="heading-one"
        >
          <div
            aria-selected="false"
            className="w-combobox__option"
            id="downshift-1-item-2"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            role="option"
          >
            <div
              className="w-combobox__option-icon"
            >
              <Icon
                name="custom"
                viewBox="0 0 1024 1024"
              >
                <path
                  d="M 83.625 "
                  key="M 83.625 "
                />
                <path
                  d="L 232.535156 "
                  key="L 232.535156 "
                />
              </Icon>
            </div>
            <div
              className="w-combobox__option-text"
            >
              Heading 1
            </div>
          </div>
        </div>
        <div
          className="w-combobox__option-row w-combobox__option-row--col2"
          key="heading-two"
        >
          <div
            aria-selected="false"
            className="w-combobox__option"
            id="downshift-1-item-3"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            role="option"
          >
            <div
              className="w-combobox__option-icon"
            >
              <span>
                H2
              </span>
            </div>
            <div
              className="w-combobox__option-text"
            >
              <span
                className="custom-text"
              >
                H2
              </span>
            </div>
          </div>
        </div>
        <div
          className="w-combobox__option-row w-combobox__option-row--col2"
          key="link"
        >
          <div
            aria-selected="false"
            className="w-combobox__option"
            id="downshift-1-item-4"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            role="option"
          >
            <div
              className="w-combobox__option-icon"
            >
              <span>
                ðŸ”—
              </span>
            </div>
            <div
              className="w-combobox__option-text"
            >
              Link
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
